File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) return true;
        
        ListNode slow = head, fast = head;

        while (fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        
        ListNode midNode = slow, curr = midNode, prev = null, next;
        
        while (curr != null){
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        
        ListNode left = head;
        ListNode right = prev;

        while (right != null){
            if (right.val != left.val){
                return false;
            }
            right = right.next;
            left = left.next;
        }
        return true;
    }
   
}

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if(!head -> next) return true;

        ListNode* slow = head;
        ListNode* fast = head;
        ListNode* dummy = NULL;
        ListNode* temp = NULL;

        while(fast && fast -> next){
            fast = fast -> next -> next;
            temp = slow -> next;
            slow -> next = dummy;
            dummy = slow;
            slow = temp;
        }
        if(fast && !fast -> next) slow = slow -> next;
        while(slow){
            if(slow -> val != dummy -> val) return false;
            slow = slow -> next;
            dummy = dummy -> next;
        }
        return true;
    }
};